src/lib/TopLevel.hs

```diff
@@ -353,26 +387,35 @@ _evalMLIR block' = do
 _evalMLIR = error "Dex built without support for MLIR"
 #endif
 
+-- TODO: there's no guarantee this name isn't already taken. Need a better story
+-- for C/LLVM function names
+mainFuncName :: SourceName
+mainFuncName = "entryFun"
+
 evalLLVM :: (Mut n, MonadPasses m) => Block n -> m n (Atom n)
 evalLLVM block = do
   backend <- backendName <$> getConfig
   bench   <- requireBenchmark
-  logger  <- getLogger
   (blockAbs, ptrVals) <- abstractPtrLiterals block
   let (cc, needsSync) = case backend of LLVMCUDA -> (EntryFun CUDARequired   , True )
                                         _        -> (EntryFun CUDANotRequired, False)
   ImpFunctionWithRecon impFun reconAtom <- checkPass ImpPass $
                                              toImpFunction backend cc blockAbs
-  LiftE llvmAST <- liftImmut do
-    DB env <- getDB
-    liftM LiftE $ liftIO $ impToLLVM env logger impFun
+  (_, llvmAST) <- impToLLVM mainFuncName impFun
   let IFunType _ _ resultTypes = impFunType impFun
   let llvmEvaluate = if bench then compileAndBench needsSync else compileAndEval
-  resultVals <- liftIO $ llvmEvaluate logger llvmAST mainFuncName ptrVals resultTypes
+  logger  <- getLogger
+  objFileNames <- liftImmut $ getObjFiles <$> getEnv
+  objFiles <- forM (eMapToList objFileNames) \(objFileName, UnitE) -> do
+    ObjectFileBinding objFile <- lookupEnv objFileName
+    return objFile
```

Comment:
This looks very expensive! We end up loading all of the precompiled object files for every single evaluation, even if it's just 2 + 2